src/ngraph_assign_clusters.cc

```diff
@@ -195,20 +240,95 @@ Status AssignClusters(Graph* graph) {
+      if (src_predicate != TRUE_PRED_STRING &&
+          dst_predicate != TRUE_PRED_STRING && src_predicate != dst_predicate) {
+        continue;
+      }
+
+      // Case src X , dst True // invalid scenario
+      if (src_predicate != TRUE_PRED_STRING &&
+          dst_predicate == TRUE_PRED_STRING) {
+        return errors::Internal("Attempting to cluster control-flow node ",
+                                dst->name(), "[", dst->type_string(), "]");
+      }
+
+      // Case src True, dst Y
+      // If this edge is contracted, all the outputs of the merged cluster will
+      // have the predicate Y. Hence contraction is possible only when, all
+      // outputs of the src cluster (other than the current edge) have the
+      // predicate Y
+      if (src_predicate == TRUE_PRED_STRING) {
+        auto src_cluster_out_edges = cluster_map[src]->outgoing_edges;
+        bool found_same_out_preds = true;
+        std::string pred_check = dst_predicate;
+
+        for (const Edge* src_cluster_edge : src_cluster_out_edges) {
+          if (src_cluster_edge == edge) {
+            continue;
+          }
+
+          Node* src_cluster_dst = src_cluster_edge->dst();
+          std::string src_cluster_dst_pred =
+              cluster_map[src_cluster_dst]->predicate_string;
+          if (pred_check != src_cluster_dst_pred) {
+            found_same_out_preds = false;
+            break;
+          }
+        }
+
+        // Cannot contract this edge
+        if (!found_same_out_preds) {
+          continue;
+        }
+      }
+
+      // Can be clustered
+      // Case src X, dst Y, X==Y
+      // Case src True, dst Y(Y can be True), all other output edges from src
+      // cluster have pred Y
+
+      // Try clustering
       if (gc.HasEdge(src_index, dst_index) &&
           gc.ContractEdge(src_index, dst_index)) {
         NGRAPH_VLOG(5) << "Contracting: " << src->name() << "["
-                       << edge->src_output() << "]@" << src_index << " -> "
-                       << dst->name() << "[" << edge->dst_input() << "]@"
-                       << dst_index;
+                       << src->type_string() << " , " << edge->src_output()
+                       << "]@" << src_index << " -> " << dst->name() << "["
+                       << dst->type_string() << " , " << edge->dst_input()
+                       << "]@" << dst_index;
+        NGRAPH_VLOG(5) << "Src pred: " << src_predicate
+                       << " ,Dst pred: " << dst_predicate;
+
+        // Merge dst cluster into src cluster
+        std::string cluster_pred =
+            (src_predicate != TRUE_PRED_STRING) ? src_predicate : dst_predicate;
+        cluster_map[src]->predicate_string = cluster_pred;
+
+        auto cluster_dst = cluster_map[dst];
+        // Update outgoing edges of the merged cluster
+        for (auto cluster_dst_out_edge : cluster_dst->outgoing_edges) {
+          cluster_map[src]->outgoing_edges.insert(cluster_dst_out_edge);
+        }
```

Comment:
this I believe is an union right?
I think it can be done without looping by passing in a range: `cluster_map[src]->outgoing_edges.insert(cluster_map[dst]->outgoing_edges.begin(), cluster_map[dst]->outgoing_edges.end())`

(but i m not sure, only the compiler (and unit tests) can tell)